README.md
# Liferay Portal CE 7, 7.1 and 7.2 Database All In One Support
[![Antonio Musarra's Blog](https://img.shields.io/badge/maintainer-Antonio_Musarra's_Blog-purple.svg?colorB=6e60cc)](https://www.dontesta.it) 
[![Twitter Follow](https://img.shields.io/twitter/follow/antonio_musarra.svg?style=social&label=%40antonio_musarra%20on%20Twitter&style=plastic)](https://twitter.com/antonio_musarra) ![travis ci](https://travis-ci.org/amusarra/liferay-portal-database-all-in-one-support.svg?branch=master)

Those who follow Liferay is aware of the fact that the Community Edition 
version 7 of Liferay, were eliminated quite a bit of components App Server, 
Database & Clustering Support. For more detail information you can read the 
blog post by [Bryan Cheung]( https://www.liferay.com/it/web/bryan.cheung/blog/-/blogs/liferay-portal-7-ce-app-server-database-clustering-support) published on April 7, 2016.



The Liferay 7 CE no more support OOTB (Out Of The Box):

* Application Server: Oracle WebLogic, IBM WebSphere
* Clustering
* MultiVM Cache
* Oracle Database, Microsoft SQL Server, IBM DB2, Sybase DB



This project add support to the [Oracle Database](https://www.oracle.com/database/), [SQL Server](https://www.microsoft.com/sql-server/sql-server-2019) and [IBM DB2](https://www.ibm.com/products/db2-database) database. 
Liferay has performed refactorting the code so that it is possible and easy 
to add support for databases no longer supported OOTB (out-of-the-box)

**Attention update:** The last version (1.1.4) of the driver works with the 
Liferay 7.2.1 CE GA2.

[<img src="https://www.dontesta.it/wp-content/uploads/2017/04/PayPalMeAntonioMusarra.png">](https://paypal.me/AntonioMusarra)

I invite you to read the article [How to build a Docker Liferay 7.2 image with the Oracle Database support](https://www.dontesta.it/en/2019/08/21/how-to-build-a-docker-liferay-7-2-image-with-the-oracle-database-support/) which 
may be interesting for you. 

In the following video, I will guide you step-by-step instructions on how to 
add support for Oracle Database to Liferay 7 Community Edition in the bundled 
version of **Wildfly**.

[![Liferay 7 Wildfly: How to add support for Oracle DB ](https://img.youtube.com/vi/7fojCjko7Ac/0.jpg)](https://www.youtube.com/watch?v=7fojCjko7Ac)

## 1. Introduction
To extend support to other databases, Liferay has decided to refactory code to 
use Java [*SPI (Service Provider Interface)*](https://docs.oracle.com/javase/tutorial/sound/SPI-intro.html). 
**SPI** is the mechanism that allows you to extend/change the behavior within a 
system without changing the source. It includes interfaces, classes or methods 
that the user extends or implements in order to obtain a certain functionality.

In short we must:
* Implement the SPI interface [com.liferay.portal.kernel.dao.db.DBFactory](https://github.com/liferay/liferay-portal/blob/7.2.1-ga2/portal-kernel/src/com/liferay/portal/kernel/dao/db/DBFactory.java)
* Implement the abstract class [com.liferay.portal.dao.db.BaseDB](https://github.com/liferay/liferay-portal/blob/7.2.1-ga2/portal-impl/src/com/liferay/portal/dao/db/BaseDB.java)

The following code shows how service providers are loaded via SPI.



```java
public DBManagerImpl() {
  ServiceLoader<DBFactory> serviceLoader = ServiceLoader.load(
    DBFactory.class, DBManagerImpl.class.getClassLoader());

  for (DBFactory dbFactory : serviceLoader) {
    _dbFactories.put(dbFactory.getDBType(), dbFactory);
  }
}
```
Source Code 1 - Shows how service providers are loaded via SPI



To register your service provider, you create a provider configuration file, 
which is stored in the **META-INF/services** directory of the service provider's 
JAR file. The name of the configuration file is the fully qualified class name 
of the service provider, in which each component of the name is separated by a 
period (.), and nested classes are separated by a dollar sign ($).

The provider configuration file contains the fully qualified class names (FQCN) 
of your service providers, one name per line. The file must be UTF-8 encoded. 
Additionally, you can include comments in the file by beginning the comment line 
with the number sign (#).

Our file (inside `META-INF/services` directory) is called `com.liferay.portal.kernel.dao.db.DBFactory` and contain the 
FQCN of this class:

1. [it.dontesta.labs.liferay.portal.dao.db.OracleDBFactory](https://github.com/amusarra/liferay-portal-database-all-in-one-support/blob/master/src/main/java/it/dontesta/labs/liferay/portal/dao/db/OracleDBFactory.java)
2. [it.dontesta.labs.liferay.portal.dao.db.SQLServerDBFactory](https://github.com/amusarra/liferay-portal-database-all-in-one-support/blob/master/src/main/java/it/dontesta/labs/liferay/portal/dao/db/SQLServerDBFactory.java)
3. [it.dontesta.labs.liferay.portal.dao.db.DB2DBFactory](https://github.com/amusarra/liferay-portal-database-all-in-one-support/blob/master/src/main/java/it/dontesta/labs/liferay/portal/dao/db/DB2DBFactory.java)



The class diagrams of each database driver are shown below.



<img src="docs/images/ClassDiagram_SQLServer_1.png" alt="ClassDiagram_SQLServer_1" style="zoom: 67%;" />

Figure 1 - Class diagram of the SQLServerDB driver



<img src="docs/images/ClassDiagram_DB2DB_1.png" alt="ClassDiagram_DB2DB_1" style="zoom: 67%;" />

Figure 2 - Class diagram of the DB2DB driver



<img src="docs/images/ClassDiagram_OracleDB_1.png" alt="ClassDiagram_OracleDB_1" style="zoom:67%;" />


Figure 3 - Class diagram of the OracleDB driver




## 2. How-To Build the project from sources
Requirements for build the project
1. Sun/Oracle JDK 1.8
2. Maven 3.x (for build project)

The driver that adds support for *Oracle*, SQLServer and *DB2* database 
is a jar artifact (**liferay-portal-database-all-in-one-support-${version}.jar**) which 
then will be installed in `ROOT/WEB-INF/lib` (for the Apache Tomcat).

To generate the all in one driver just follow the instructions below.

**You can download the latest version binary jar from Maven Central Repository**
[liferay-portal-database-all-in-one-support](https://search.maven.org/#search%7Cga%7C1%7Cit.dontesta), 
by doing so you can avoid doing the build.



```bash
$ git clone https://github.com/amusarra/liferay-portal-database-all-in-one-support.git
$ cd liferay-portal-database-all-in-one-support
$ mvn package
```

Console 1 - Clone the source project from GitHub and run package goal



The build process create the jar 
`liferay-portal-database-all-in-one-support-${version}.jar` inside the (maven) 
target directory.

## 3. How-To Configure Liferay Portal

Below you can see the `portal-ext.properties`. In the sample file are shown JDBC 
configurations sample for Oracle, SQL Server, and DB2.



```properties
##
## JDBC
##

    #
    # Oracle
    #
    # jdbc.default.driverClassName=oracle.jdbc.OracleDriver
    # jdbc.default.username=liferayce7
    # jdbc.default.password=liferay12345
    # jdbc.default.url=jdbc:oracle:thin:@oracledb.vm.local:1521:xe

    #
    # DB2
    #
    # jdbc.default.driverClassName=com.ibm.db2.jcc.DB2Driver
    # jdbc.default.url=jdbc:db2://db2.vm.local:50001/lportal:deferPrepares=false;fullyMaterializeInputStreams=true;fullyMaterializeLobData=true;progresssiveLocators=2;progressiveStreaming=2;
    # jdbc.default.username=db2inst1
    # jdbc.default.password=system

    #
    # SQL Server
    #
    # jdbc.default.driverClassName=com.microsoft.sqlserver.jdbc.SQLServerDriver
    # jdbc.default.username=liferay
    # jdbc.default.password=liferay12345
    # jdbc.default.url=jdbc:sqlserver://sqlserverdb.vm.local;databaseName=liferayce7
```

Source Code 2 - Sample JDBC connection strings for Oracle, SQL Server and DB2 database



You could also configure database access as a **[JNDI](https://en.wikipedia.org/wiki/Java_Naming_and_Directory_Interface)** resource and specify the 
resource name in configuration.

```properties
##
## JDBC
##

		#
		# Set the JNDI name to lookup the JDBC data source. If none is set,
		# then the portal will attempt to create the JDBC data source based on the
		# properties prefixed with "jdbc.default.".
		#
		jdbc.default.jndi.name=java:jdbc/LiferayPool
```

Source Code 3 - Configure the JDBC connection via JNDI



In order for Liferay to be able to connect to the database, it is necessary to 
install the JDBC driver compatible with the version of the specific database and 
JVM. Here are the links to the resources to download the JDBC driver.

1. [Oracle JDBC Drivers](https://www.oracle.com/database/technologies/appdev/jdbc-downloads.html)
2. [Microsoft SQL Server JDBC Drivers and support matrix](https://docs.microsoft.com/en-us/sql/connect/jdbc/microsoft-jdbc-driver-for-sql-server-support-matrix?view=sql-server-ver15)
3. [IBM DB2 Database](https://www.ibm.com/support/pages/db2-jdbc-driver-versions-and-downloads)



The following documents (see database section) provide details of the 
configurations that are certified by Liferay. You can see the complete 
documents on [Liferay Portal](https://web.liferay.com/it/services/support/compatibility-matrix).

![Liferay_72_Compatibility_Matrix_Database](docs/images/Liferay_72_Compatibility_Matrix_Database.png)

## 4. Other useful resources



1. [How to build a Docker Liferay 7.2 image with the Oracle Database support](https://www.dontesta.it/en/2019/08/21/how-to-build-a-docker-liferay-7-2-image-with-the-oracle-database-support/)
2. [How to build a Docker Liferay 7.2 image with the SQL Server 2017 Database support](https://www.dontesta.it/en/2019/10/06/how-to-build-a-docker-liferay-7-2-image-with-the-sql-server-2017-database-support/)
3. [Liferay 7.1: How to add support for Oracle DB](https://www.dontesta.it/en/2018/10/07/liferay-7-1-how-to-add-support-for-oracle-db/)
4. [Come installare Liferay 7 su JBOSS EAP con il supporto Oracle Database](https://www.slideshare.net/amusarra/come-installare-liferay-7-su-jboss-eap-con-il-support-oracle-database) 
5. [Liferay 7 Wildfly: How to add support for Oracle DB](https://www.youtube.com/watch?v=7fojCjko7Ac) (video on Antonio Musarra's Blog YouTube Channel)
6. [Liferay 7 Community Edition GA5 & Oracle 12c via Docker Composer](https://www.youtube.com/watch?v=yLVCEl8L8cU) (video on Antonio Musarra's Blog YouTube Channel)
7. [Come installare Liferay 7 su JBoss EAP con il supporto per Oracle Database](https://www.youtube.com/watch?v=QaVaP89yWiM&t=848s) (video on Antonio Musarra's Blog YouTube Channel)

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [1.1.4] - 2020-03-03
### Added
- Added support for Liferay 7.2.1 GA2

## [1.1.3] - 2019-06-26
### Added
- Added support for Liferay 7.2 GA1

## [1.1.2] - 2019-02-02
### Added
- Added support for Liferay 7.1 GA3

## [1.1.1] - 2019-02-02
### Added
- Added support for Liferay 7.1 GA2

## [1.1.0] - 2018-10-04
### Added
- Added support for Liferay 7.1 GA1

## [1.0.0] - 2017-03-27
- First release for Liferay 7.1

src/test/java/it/dontesta/labs/liferay/portal/db/BD2DBTest.java
package it.dontesta.labs.liferay.portal.db;

import com.liferay.portal.kernel.dao.db.DB;
import com.liferay.portal.kernel.dao.db.DBType;

import it.dontesta.labs.liferay.portal.dao.db.DB2DB;

import java.io.IOException;

import org.junit.Assert;
import org.junit.Test;

/**
 * @author Antonio Musarra
 */
public class BD2DBTest extends BaseDBTestCase {

	@Test
	public void testDBType() throws IOException {
		Assert.assertEquals(
				DBType.DB2,
				getDB().getDBType());
	}

	@Override
	protected DB getDB() {
		return new DB2DB(0, 0);
	}

}

src/main/java/it/dontesta/labs/liferay/portal/dao/db/DB2DB.java
/**
 * Copyright (c) 2000-2013 Liferay, Inc. All rights reserved.
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

package it.dontesta.labs.liferay.portal.dao.db;

import java.io.IOException;
import java.sql.CallableStatement;
import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Types;
import java.util.HashSet;
import java.util.Set;

import com.liferay.portal.dao.db.BaseDB;
import com.liferay.portal.kernel.dao.db.DBType;
import com.liferay.portal.kernel.dao.jdbc.DataAccess;
import com.liferay.portal.kernel.io.unsync.UnsyncBufferedReader;
import com.liferay.portal.kernel.io.unsync.UnsyncStringReader;
import com.liferay.portal.kernel.util.CharPool;
import com.liferay.portal.kernel.util.StringBundler;
import com.liferay.portal.kernel.util.StringUtil;

/**
 * @author Alexander Chow
 * @author Sandeep Soni
 * @author Ganesh Ram
 * @author Bruno Farache
 * @author Antonio Musarra
 * @author Javier Alpanez
 */
public class DB2DB extends BaseDB {

	public DB2DB(int majorVersion, int minorVersion) {
		super(DBType.DB2, majorVersion, minorVersion);
	}

	@Override
	public String buildSQL(String template) throws IOException {
		template = convertTimestamp(template);
		template = replaceTemplate(template, getTemplate());

		template = reword(template);
		template = removeLongInserts(template);
		template = removeNull(template);
		template = StringUtil.replace(template, "\\'", "''");
		template = StringUtil.replace(template, "\\n", "'||CHR(10)||'");

		return template;
	}

	@Override
	public boolean isSupportsAlterColumnType() {
		return _SUPPORTS_ALTER_COLUMN_TYPE;
	}

	@Override
	public boolean isSupportsInlineDistinct() {
		return _SUPPORTS_INLINE_DISTINCT;
	}

	@Override
	public boolean isSupportsScrollableResults() {
		return _SUPPORTS_SCROLLABLE_RESULTS;
	}

	@Override
	public void runSQL(String template) throws IOException, SQLException {
		if (template.startsWith(ALTER_COLUMN_NAME)) {
			String sql = buildSQL(template);

			String[] alterSqls = StringUtil.split(sql, CharPool.SEMICOLON);

			for (String alterSql : alterSqls) {
				alterSql = StringUtil.trim(alterSql);

				runSQL(alterSql);
			}
		}
		else {
			super.runSQL(template);
		}
	}

	@Override
	public void runSQL(String[] templates) throws IOException, SQLException {
		super.runSQL(templates);

		reorgTables(templates);
	}

	@Override
	protected String buildCreateFileContent(
			String sqlDir, String databaseName, int population)
		throws IOException {

		String suffix = getSuffix(population);

		StringBundler sb = new StringBundler(15);

		sb.append("drop database ");
		sb.append(databaseName);
		sb.append(";\n");
		sb.append("create database ");
		sb.append(databaseName);
		sb.append(" pagesize 32768 temporary tablespace managed by automatic ");
		sb.append("storage;\n");

		if (population != BARE) {
			sb.append("connect to ");
			sb.append(databaseName);
			sb.append(";\n");
			sb.append(getCreateTablesContent(sqlDir, suffix));
			sb.append("\n\n");
			sb.append(readFile(sqlDir + "/indexes/indexes-db2.sql"));
			sb.append("\n\n");
			sb.append(readFile(sqlDir + "/sequences/sequences-db2.sql"));
		}

		return sb.toString();
	}

	@Override
	protected String getServerName() {
		return "db2";
	}

	@Override
	protected int[] getSQLTypes() {
		return _SQL_TYPES;
	}
	
	@Override
	protected String[] getTemplate() {
		return _DB2;
	}
	
	protected boolean isRequiresReorgTable(Connection con, String tableName)
		throws SQLException {

		boolean reorgTableRequired = false;

		PreparedStatement ps = null;
		ResultSet rs = null;

		try {
			StringBundler sb = new StringBundler(4);

			sb.append("select num_reorg_rec_alters from table(");
			sb.append("sysproc.admin_get_tab_info(current_schema, '");
			sb.append(StringUtil.toUpperCase(tableName));
			sb.append("')) where reorg_pending = 'Y'");

			ps = con.prepareStatement(sb.toString());

			rs = ps.executeQuery();

			if (rs.next()) {
				int numReorgRecAlters = rs.getInt(1);

				if (numReorgRecAlters >= 1) {
					reorgTableRequired = true;
				}
			}
		}
		finally {
			DataAccess.cleanUp(ps, rs);
		}

		return reorgTableRequired;
	}

	protected void reorgTable(Connection con, String tableName)
		throws SQLException {

		if (!isRequiresReorgTable(con, tableName)) {
			return;
		}

		CallableStatement callableStatement = null;

		try {
			callableStatement = con.prepareCall("call sysproc.admin_cmd(?)");

			callableStatement.setString(1, "reorg table " + tableName);

			callableStatement.execute();
		}
		finally {
			DataAccess.cleanUp(callableStatement);
		}
	}

	protected void reorgTables(String[] templates) throws SQLException {
		Set<String> tableNames = new HashSet<>();

		for (String template : templates) {
			if (template.startsWith("alter table")) {
				tableNames.add(template.split(" ")[2]);
			}
		}

		if (tableNames.isEmpty()) {
			return;
		}

		Connection con = null;

		try {
			con = DataAccess.getConnection();

			for (String tableName : tableNames) {
				reorgTable(con, tableName);
			}
		}
		finally {
			DataAccess.cleanUp(con);
		}
	}

	@Override
	protected String reword(String data) throws IOException {
		try (UnsyncBufferedReader unsyncBufferedReader =
				new UnsyncBufferedReader(new UnsyncStringReader(data))) {

			StringBundler sb = new StringBundler();

			String line = null;

			while ((line = unsyncBufferedReader.readLine()) != null) {
				if (line.startsWith(ALTER_COLUMN_NAME)) {
					String[] template = buildColumnNameTokens(line);

					line = StringUtil.replace(
						"alter table @table@ add column @new-column@ @type@;\n",
						REWORD_TEMPLATE, template);
					line += StringUtil.replace(
						"update @table@ set @new-column@ = @old-column@;\n",
						REWORD_TEMPLATE, template);
					line += StringUtil.replace(
						"alter table @table@ drop column @old-column@",
						REWORD_TEMPLATE, template);
				}
				else if (line.startsWith(ALTER_TABLE_NAME)) {
					String[] template = buildTableNameTokens(line);

					line = StringUtil.replace(
						"alter table @old-table@ to @new-table@;",
						RENAME_TABLE_TEMPLATE, template);
				}
				else if (line.contains(DROP_INDEX)) {
					String[] tokens = StringUtil.split(line, ' ');

					line = StringUtil.replace(
						"drop index @index@;", "@index@", tokens[2]);
				}

				sb.append(line);
				sb.append("\n");
			}

			return sb.toString();
		}
	}

	private static final String[] _DB2 = {
		"--", "1", "0", "'1970-01-01-00.00.00.000000'", "current timestamp",
		" blob", " blob", " smallint", " timestamp", " double", " integer",
		" bigint", " varchar(4000)", " clob", " varchar",
		" generated always as identity", "commit"
	};
	
	private static final int[] _SQL_TYPES = {
			Types.BLOB, Types.BLOB, Types.SMALLINT, Types.TIMESTAMP, Types.DOUBLE,
			Types.INTEGER, Types.BIGINT, Types.VARCHAR, Types.CLOB, Types.VARCHAR
	};

	private static final boolean _SUPPORTS_ALTER_COLUMN_TYPE = false;

	private static final boolean _SUPPORTS_INLINE_DISTINCT = false;

	private static final boolean _SUPPORTS_SCROLLABLE_RESULTS = false;
}


src/main/java/it/dontesta/labs/liferay/portal/dao/db/DB2DBFactory.java
/**
 * Copyright (c) 2000-2013 Liferay, Inc. All rights reserved.
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

package it.dontesta.labs.liferay.portal.dao.db;

import com.liferay.portal.kernel.dao.db.BaseDBFactory;
import com.liferay.portal.kernel.dao.db.DB;
import com.liferay.portal.kernel.dao.db.DBType;

/**
 * @author Antonio Musarra
 */
public class DB2DBFactory extends BaseDBFactory {

	@Override
	public DB doCreate(int dbMajorVersion, int dbMinorVersion) {
		return new DB2DB(dbMajorVersion, dbMinorVersion);
	}

	@Override
	public DBType getDBType() {
		return DBType.DB2;
	}
}


src/main/java/it/dontesta/labs/liferay/portal/dao/db/OracleDBFactory.java
/**
 * Copyright (c) 2000-2013 Liferay, Inc. All rights reserved.
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

package it.dontesta.labs.liferay.portal.dao.db;

import com.liferay.portal.kernel.dao.db.BaseDBFactory;
import com.liferay.portal.kernel.dao.db.DB;
import com.liferay.portal.kernel.dao.db.DBType;

/**
 * @author Antonio Musarra
 */
public class OracleDBFactory extends BaseDBFactory {

	@Override
	public DB doCreate(int dbMajorVersion, int dbMinorVersion) {
		return new OracleDB(dbMajorVersion, dbMinorVersion);
	}

	@Override
	public DBType getDBType() {
		return DBType.ORACLE;
	}
}


src/main/java/it/dontesta/labs/liferay/portal/dao/db/SQLServerDBFactory.java
/**
 * Copyright (c) 2000-2013 Liferay, Inc. All rights reserved.
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

package it.dontesta.labs.liferay.portal.dao.db;

import com.liferay.portal.kernel.dao.db.BaseDBFactory;
import com.liferay.portal.kernel.dao.db.DB;
import com.liferay.portal.kernel.dao.db.DBType;

/**
 * @author Antonio Musarra
 */
 public class SQLServerDBFactory extends BaseDBFactory {

 	@Override
 	public DB doCreate(int dbMajorVersion, int dbMinorVersion) {
 		return new SQLServerDB(dbMajorVersion, dbMinorVersion);
 	}

 	@Override
 	public DBType getDBType() {
 		return DBType.SQLSERVER;
 	}

 }


src/main/java/it/dontesta/labs/liferay/portal/dao/db/SQLServerDB.java
/**
 * Copyright (c) 2000-2013 Liferay, Inc. All rights reserved.
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

package it.dontesta.labs.liferay.portal.dao.db;

import java.io.IOException;
import java.sql.Connection;
import java.sql.DatabaseMetaData;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Types;
import java.util.ArrayList;
import java.util.List;

import com.liferay.portal.dao.db.BaseDB;
import com.liferay.portal.kernel.dao.db.DBType;
import com.liferay.portal.kernel.dao.db.Index;
import com.liferay.portal.kernel.dao.jdbc.DataAccess;
import com.liferay.portal.kernel.io.unsync.UnsyncBufferedReader;
import com.liferay.portal.kernel.io.unsync.UnsyncStringReader;
import com.liferay.portal.kernel.util.StringBundler;
import com.liferay.portal.kernel.util.StringPool;
import com.liferay.portal.kernel.util.StringUtil;

/**
 * @author Alexander Chow
 * @author Sandeep Soni
 * @author Ganesh Ram
 * @author Antonio Musarra
 * @author Javier Alpanez
 */
public class SQLServerDB extends BaseDB {

	public SQLServerDB(int majorVersion, int minorVersion) {
		super(DBType.SQLSERVER, majorVersion, minorVersion);
	}

	@Override
		public String buildSQL(String template) throws IOException {
			template = convertTimestamp(template);
			template = replaceTemplate(template, getTemplate());

			template = reword(template);
			template = StringUtil.replace(template, "\ngo;\n", "\ngo\n");
			template = StringUtil.replace(
				template,
				new String[] {"\\\\", "\\'", "\\\"", "\\n", "\\r"},
				new String[] {"\\", "''", "\"", "\n", "\r"});

			return template;
		}

	@Override
	public List<Index> getIndexes(Connection con) throws SQLException {
		List<Index> indexes = new ArrayList<>();

			PreparedStatement ps = null;
			ResultSet rs = null;

			try {
				DatabaseMetaData databaseMetaData = con.getMetaData();

				if (databaseMetaData.getDatabaseMajorVersion() <=
						_SQL_SERVER_2000) {

					return indexes;
				}

				StringBundler sb = new StringBundler(6);

				sb.append("select sys.tables.name as table_name, ");
				sb.append("sys.indexes.name as index_name, is_unique from ");
				sb.append("sys.indexes inner join sys.tables on ");
				sb.append("sys.tables.object_id = sys.indexes.object_id where ");
				sb.append("sys.indexes.name like 'LIFERAY_%' or sys.indexes.name ");
				sb.append("like 'IX_%'");

				String sql = sb.toString();

				ps = con.prepareStatement(sql);

				rs = ps.executeQuery();

				while (rs.next()) {
					String indexName = rs.getString("index_name");
					String tableName = rs.getString("table_name");
					boolean unique = !rs.getBoolean("is_unique");

				indexes.add(new Index(indexName, tableName, unique));
			}
		}
		finally {
			DataAccess.cleanUp(ps, rs);
		}

			return indexes;
		}

		@Override
		public boolean isSupportsAlterColumnType() {
			return _SUPPORTS_ALTER_COLUMN_TYPE;
		}

		@Override
			protected String buildCreateFileContent(
					String sqlDir, String databaseName, int population)
				throws IOException {

				String suffix = getSuffix(population);

				StringBundler sb = new StringBundler(17);

				sb.append("drop database ");
				sb.append(databaseName);
				sb.append(";\n");
				sb.append("create database ");
				sb.append(databaseName);
				sb.append(";\n");
				sb.append("\n");
				sb.append("go\n");
				sb.append("\n");

				if (population != BARE) {
					sb.append("use ");
					sb.append(databaseName);
					sb.append(";\n\n");
					sb.append(getCreateTablesContent(sqlDir, suffix));
					sb.append("\n\n");
					sb.append(readFile(sqlDir + "/indexes/indexes-sql-server.sql"));
					sb.append("\n\n");
					sb.append(readFile(sqlDir + "/sequences/sequences-sql-server.sql"));
				}

				return sb.toString();
			}

			@Override
			protected String getServerName() {
				return "sql-server";
			}

			@Override
			protected int[] getSQLTypes() {
				return _SQL_TYPES;
			}
			
			@Override
			protected String[] getTemplate() {
				return _SQL_SERVER;
			}

	@Override
	protected String reword(String data) throws IOException {
		try (UnsyncBufferedReader unsyncBufferedReader =
				new UnsyncBufferedReader(new UnsyncStringReader(data))) {

				StringBundler sb = new StringBundler();

				String line = null;

				while ((line = unsyncBufferedReader.readLine()) != null) {
					if (line.startsWith(ALTER_COLUMN_NAME)) {
						String[] template = buildColumnNameTokens(line);

					line = StringUtil.replace(
						"exec sp_rename '@table@.@old-column@', " +
							"'@new-column@', 'column';",
						REWORD_TEMPLATE, template);
				}
				else if (line.startsWith(ALTER_COLUMN_TYPE)) {
					String[] template = buildColumnTypeTokens(line);

						line = StringUtil.replace(
							"alter table @table@ alter column @old-column@ @type@;",
							REWORD_TEMPLATE, template);
					}
					else if (line.startsWith(ALTER_TABLE_NAME)) {
						String[] template = buildTableNameTokens(line);

						line = StringUtil.replace(
							"exec sp_rename '@old-table@', '@new-table@';",
							RENAME_TABLE_TEMPLATE, template);
					}
					else if (line.contains(DROP_INDEX)) {
						String[] tokens = StringUtil.split(line, ' ');

						String tableName = tokens[4];

					if (tableName.endsWith(StringPool.SEMICOLON)) {
						tableName = tableName.substring(
							0, tableName.length() - 1);
					}

						line = StringUtil.replace(
							"drop index @table@.@index@;", "@table@", tableName);
						line = StringUtil.replace(line, "@index@", tokens[2]);
					}

					sb.append(line);
					sb.append("\n");
				}

			return sb.toString();
		}
	}

			private static final String[] _SQL_SERVER = {
				"--", "1", "0", "'19700101'", "GetDate()", " image", " image", " bit",
				" datetime", " float", " int", " bigint", " nvarchar(4000)",
				" nvarchar(max)", " nvarchar", "  identity(1,1)", "go"
			};

			private static final int _SQL_SERVER_2000 = 8;
			
			private static final int[] _SQL_TYPES = {
					Types.LONGVARBINARY, Types.LONGVARBINARY, Types.BIT, Types.TIMESTAMP,
					Types.DOUBLE, Types.INTEGER, Types.BIGINT, Types.NVARCHAR,
					Types.NVARCHAR, Types.NVARCHAR
			};
			
			private static final boolean _SUPPORTS_ALTER_COLUMN_TYPE = false;
}
